nifi-nar-bundles/nifi-protobuf-bundle/nifi-protobuf-services/src/main/java/org/apache/nifi/services/protobuf/converter/ProtobufDataConverter.java

```diff
@@ -0,0 +1,380 @@
+        }
+
+        for (final OneOf oneOf : messageType.getOneOfs()) {
+            for (Field field : oneOf.getFields()) {
+                getField(new ProtoField(field), unknownFieldSet.getField(field.getTag()), recordValues);
+            }
+        }
+        return recordValues;
+    }
+
+    /**
+     * Checks the field value's presence and sets it into the result Map.
+     *
+     * @param protoField   proto field's properties
+     * @param unknownField field's value
+     * @param values       Map of values
+     */
+    private void getField(ProtoField protoField, UnknownFieldSet.Field unknownField, Map<String, Object> values) throws InvalidProtocolBufferException {
+        Optional<Object> fieldValue = convertFieldValues(protoField, unknownField);
+        fieldValue.ifPresent(o -> values.put(protoField.getFieldName(), o));
+    }
+
+    private Optional<Object> convertFieldValues(ProtoField protoField, UnknownFieldSet.Field unknownField) throws InvalidProtocolBufferException {
+        if (!unknownField.getLengthDelimitedList().isEmpty()) {
+            return Optional.of(convertLengthDelimitedFields(protoField, unknownField.getLengthDelimitedList()));
+        }
+        if (!unknownField.getFixed32List().isEmpty()) {
+            return Optional.of(convertFixed32Fields(protoField, unknownField.getFixed32List()));
+        }
+        if (!unknownField.getFixed64List().isEmpty()) {
+            return Optional.of(convertFixed64Fields(protoField, unknownField.getFixed64List()));
+        }
+        if (!unknownField.getVarintList().isEmpty()) {
+            return Optional.of(convertVarintFields(protoField, unknownField.getVarintList()));
+        }
+
+        return Optional.empty();
+    }
+
+    /**
+     * Converts a Length-Delimited field value into it's suitable data type.
+     *
+     * @param protoField proto field's properties
+     * @param values     field's unprocessed values
+     * @return converted field values
+     * @throws InvalidProtocolBufferException failed to parse input data
+     */
+    private Object convertLengthDelimitedFields(ProtoField protoField, List<ByteString> values) throws InvalidProtocolBufferException {
+        final ProtoType protoType = protoField.getProtoType();
+        if (protoType.isScalar()) {
+            switch (FieldType.findValue(protoType.getSimpleName())) {
+                case STRING:
+                    return resolveFieldValue(protoField, values, ByteString::toStringUtf8);
+                case BYTES:
+                    return resolveFieldValue(protoField, values, ByteString::toByteArray);
+                default:
+                    throw new IllegalStateException(String.format("Incompatible value was received for field [%s]," +
+                            " [%s] is not LengthDelimited field type", protoField.getFieldName(), protoType.getSimpleName()));
+            }
```

Comment:
The `FieldType` enum cannot be removed since it is being used in the `ProtoSchemaParser` for DataType matching. If I remove it I need to define the scalar proto field types in two different place which is not efficient. Instead I changed the `Enum` matching to assign only the '`valueConverter`' function so the multiple '`return`' were removed.